moveit_core/robot_state/src/robot_state.cpp

```diff
@@ -1850,7 +1818,7 @@ bool RobotState::setFromIKSubgroups(const JointModelGroup* jmg, const EigenSTL::
       std::vector<double> ik_sol;
       moveit_msgs::MoveItErrorCodes error;
       const std::vector<double>& climits = consistency_limits.empty() ? std::vector<double>() : consistency_limits[sg];
-      if (solvers[sg]->searchPositionIK(ik_queries[sg], seed, timeout, climits, ik_sol, error))
+      if (solvers[sg]->searchPositionIK(ik_queries[sg], seed, timeout / sub_groups.size(), climits, ik_sol, error))
```

Comment:
Please justify this change

Reply:
The given timeout is an overall timeout and not intended to be consumed by each sub group.

Reply:
But then you allow each sub-group to be run multiple times with `while ((ros::WallTime::now() - start).toSec() > timeout);` right? To be honest, this method is a mess and should be split up into into sub-functions so if I am missing things let me know.

Reply:
Please, don't blame me for this function. I didn't write it in first place. I just fixed an obvious issue, that allowed this function to take `N * timeout` time (instead of timeout), where N is the number of sub groups.
If you have a proposal how to improve this function, please file a separate PR. For me, it looks quite ok in general: All sub group solvers are run to find a solution for their sub groups. Usually this yields a valid solution, except a custom constraint function will reject it. In this latter case (and only then), the while loop allows another attempt with another random seed for the whole group - of course only if time is left.

Reply:
I don't blame you at all. My apologies if my comment came off that way. I only meant to convey that I don't entirely get what is happening here so my feedback might be a little mislead. I agree that this change is an improvement even though it still is not guaranteed to return before the timeout. After scanning the method quickly the biggest question I had was what happens in the case where 3 out of 4 subgroups succeed? Will the loop recompute all 4 IK solutions or just the 1?

Reply:
Thanks for your explanation, with that in mind, does it even make sense to try again on failure? If we give each subgroup `(timeout - elapsed) / (len(remaining_subgroups))` we could guarantee that the method returns in time and still make the most of the available time. It strikes me as strange to be randomly seeding here when really the kinematic solver's should be doing the random restarts.

Reply:
> Will the loop recompute all 4 IK solutions or just the 1?

I think so, yes. You are right that all sub group solvers do their seeding locally already. Hence, an IK failure of a sub group probably is fatal anyway, meaning that no solution will be found.
On the other hand, if all IK solvers do find a solution, but their combination is rejected due to feasibility (e.g. collision), it completely makes sense to globally re-seed to resolve the situation.
Even better, this function should collect _all_ potential IK solutions from sub groups and then just validate their combinations for feasibility. But there is no API available for this yet (see #1285).

> It still is not guaranteed to return before the timeout.

Why not? `timeout / N * N <= timeout`. Are you talking about consecutive attempts?
Then you are probably right, because there is no timeout checking within the loop.
However, you proposal to

> give each subgroup (timeout - elapsed) / (len(remaining_subgroups))

I don't like too much, because it greedily assigns all the remaining time to the remaining groups,
probably leaving no time at all for consecutive attempts. Better, let's compute the remaining time after each attempt and distribute this equally to all sub groups.

Reply:
> Better, let's compute the remaining time after each attempt and distribute this equally to all sub groups.

Yup. That's what I was thinking :)

Reply:
There is a subtle but important difference in our suggestions. There are two loops, an outer "attempts" loop and an inner loop over all sub groups.
You suggested to assign all the remaining time (divided by the remaining subgroups) to each sub group IK call, which greedily consumes all the available time: The last sub group IK call assigns all the remaining time to this call, thus leaving no more time for an additional outer loop.

My suggestion was to assign all the remaining time after an _outer loop_, but equally distribute this time to all sub groups within the current outer loop iteration. If early sub group solvers don't consume their assigned time, in your version, later sub group solvers get more time assigned, which I think is weird.
In my suggestion, this saved time will be available for additional outer loops, again equally distributing the remaining time to all sub group solvers.

Reply:
True. I think your suggestion makes more sense if the subgroup solutions can be invalidated by one another.